Integration test for TestCorpusStructure fails locally

**Describe the bug**

When running `mvn clean install` on https://github.com/hexatomic/hexatomic/pull/117, the build fails.

**To reproduce the bug**
Steps to reproduce the behavior:
1. `git checkout feature/zest-graph-view`
2. `git pull`
3. `mvn clean install`
4. Build fails with stack trace:

```
Running org.corpus_tools.hexatomic.it.tests.TestGraphEditor
log4j:WARN No appenders could be found for logger (org.eclipse.swtbot.swt.finder.matchers.AbstractMatcher).
log4j:WARN Please initialize the log4j system properly.
*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug

*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug


(org.corpus_tools.hexatomic:26489): Gtk-CRITICAL **: 21:07:47.690: gtk_widget_get_preferred_width_for_height: assertion 'height >= 0' failed

(org.corpus_tools.hexatomic:26489): Gtk-WARNING **: 21:07:47.690: gtk_widget_size_allocate(): attempt to allocate widget with width 12 and height -13

(org.corpus_tools.hexatomic:26489): Gtk-CRITICAL **: 21:07:47.690: gtk_box_gadget_distribute: assertion 'size >= 0' failed in GtkScrollbar
*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug

*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug

*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug

*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug

Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.532 s - in org.corpus_tools.hexatomic.it.tests.TestGraphEditor
testShowSaltExample  Time elapsed: 0.422 s
Running org.corpus_tools.hexatomic.it.tests.TestCorpusStructure
Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 10.294 s <<< FAILURE! - in org.corpus_tools.hexatomic.it.tests.TestCorpusStructure
testRenameDocument  Time elapsed: 5.273 s  <<< ERROR!
org.eclipse.swtbot.swt.finder.exceptions.WidgetNotFoundException: Timed out waiting for tree item corpus_graph_1
	at org.corpus_tools.hexatomic.it.tests.TestCorpusStructure.testRenameDocument(TestCorpusStructure.java:34)
Caused by: org.eclipse.swtbot.swt.finder.widgets.TimeoutException: Timeout after: 5000 ms.: Could not find node with text corpus_graph_1
	at org.corpus_tools.hexatomic.it.tests.TestCorpusStructure.testRenameDocument(TestCorpusStructure.java:34)

testFilter  Time elapsed: 5.018 s  <<< ERROR!
org.eclipse.swtbot.swt.finder.exceptions.WidgetNotFoundException: Could not find widget matching: (of type 'Tree')
	at org.corpus_tools.hexatomic.it.tests.TestCorpusStructure.testFilter(TestCorpusStructure.java:72)
Caused by: org.eclipse.swtbot.swt.finder.widgets.TimeoutException: Timeout after: 5000 ms.: Could not find widget matching: (of type 'Tree')
	at org.corpus_tools.hexatomic.it.tests.TestCorpusStructure.testFilter(TestCorpusStructure.java:72)


Results:

Errors: 
  TestCorpusStructure.testFilter:72 » WidgetNotFound Could not find widget match...
  TestCorpusStructure.testRenameDocument:34 » WidgetNotFound Timed out waiting f...

Tests run: 3, Failures: 0, Errors: 2, Skipped: 0

```

**Expected behavior**

Integration tests pass locally.

**Screenshots**

![image](https://user-images.githubusercontent.com/3007126/73782698-d1455280-4792-11ea-9573-05063af3ff1c.png)


**Desktop (please complete the following information):**
 - Operating System: Ubuntu 18.04
 - Hexatomic Version: https://github.com/hexatomic/hexatomic/pull/117
 - Java Version: OpenJDK 8

Right, I can reproduce the issue on Ubuntu with the following `locale` settings:

```
LANG=en_GB.UTF-8
LANGUAGE=en_GB:en
LC_CTYPE="en_GB.UTF-8"
LC_NUMERIC=de_DE.UTF-8
LC_TIME=de_DE.UTF-8
LC_COLLATE="en_GB.UTF-8"
LC_MONETARY=de_DE.UTF-8
LC_MESSAGES="en_GB.UTF-8"
LC_PAPER=de_DE.UTF-8
LC_NAME=de_DE.UTF-8
LC_ADDRESS=de_DE.UTF-8
LC_TELEPHONE=de_DE.UTF-8
LC_MEASUREMENT=de_DE.UTF-8
LC_IDENTIFICATION=de_DE.UTF-8
LC_ALL=
```

Caused by:

```log
java.lang.ExceptionInInitializerError
	at org.corpus_tools.hexatomic.it.tests.TestCorpusStructure.createExampleStructure(TestCorpusStructure.java:90)
	at org.corpus_tools.hexatomic.it.tests.TestCorpusStructure.testRenameDocument(TestCorpusStructure.java:103)
Caused by: java.lang.IllegalArgumentException: EN_GB.keyboard not found, see http://wiki.eclipse.org/SWTBot/Keyboard_Layouts for more information.
	at org.corpus_tools.hexatomic.it.tests.TestCorpusStructure.createExampleStructure(TestCorpusStructure.java:90)
	at org.corpus_tools.hexatomic.it.tests.TestCorpusStructure.testRenameDocument(TestCorpusStructure.java:103)
```

--- a/.gitignore
+++ b/.gitignore
@@ -2,3 +2,4 @@ target/
 book/
 NOTES.md
 workspace/
+tests/**/Rerun*

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -22,6 +22,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - Update views via notification even if the object that changed was created before the view.
 - Allow to run Hexatomic on Java 11 platforms
+- Set locale in UI tests to avoid issues with auto-detected keyboard layouts
 
 ## [0.3.1] - 2019-12-16
 

--- a/bundles/org.corpus_tools.hexatomic.grid/src/main/java/org/corpus_tools/hexatomic/grid/style/LabelAccumulator.java
+++ b/bundles/org.corpus_tools.hexatomic.grid/src/main/java/org/corpus_tools/hexatomic/grid/style/LabelAccumulator.java
@@ -75,7 +75,7 @@ public class LabelAccumulator extends ColumnOverrideLabelAccumulator {
 
     int columnIndex = bodyDataLayer.getColumnIndexByPosition(columnPosition);
     List<Column> columns = bodyDataProvider.getColumns();
-    if (columns.size() != 0) {
+    if (!columns.isEmpty()) {
       ColumnType columnType = columns.get(columnIndex).getColumnType();
       if (columnType == ColumnType.SPAN_ANNOTATION) {
         // Assign the span annotation style label to cells in columns with the respective flag.

--- a/bundles/org.corpus_tools.hexatomic.grid/src/main/java/org/corpus_tools/hexatomic/grid/style/LabelAccumulator.java
+++ b/bundles/org.corpus_tools.hexatomic.grid/src/main/java/org/corpus_tools/hexatomic/grid/style/LabelAccumulator.java
@@ -75,11 +75,8 @@ public class LabelAccumulator extends ColumnOverrideLabelAccumulator {
 
     int columnIndex = bodyDataLayer.getColumnIndexByPosition(columnPosition);
     List<Column> columns = bodyDataProvider.getColumns();
-    ColumnType columnType = null;
     if (columns.size() != 0) {
-      columnType = columns.get(columnIndex).getColumnType();
-    }
-    if (columnType != null) {
+      ColumnType columnType = columns.get(columnIndex).getColumnType();
       if (columnType == ColumnType.SPAN_ANNOTATION) {
         // Assign the span annotation style label to cells in columns with the respective flag.
         configLabels.addLabel(StyleConfiguration.SPAN_ANNOTATION_CELL_STYLE);

--- a/bundles/org.corpus_tools.hexatomic.grid/src/main/java/org/corpus_tools/hexatomic/grid/style/LabelAccumulator.java
+++ b/bundles/org.corpus_tools.hexatomic.grid/src/main/java/org/corpus_tools/hexatomic/grid/style/LabelAccumulator.java
@@ -21,6 +21,8 @@
 
 package org.corpus_tools.hexatomic.grid.style;
 
+import java.util.List;
+import org.corpus_tools.hexatomic.grid.data.Column;
 import org.corpus_tools.hexatomic.grid.data.Column.ColumnType;
 import org.corpus_tools.hexatomic.grid.data.GraphDataProvider;
 import org.eclipse.nebula.widgets.nattable.layer.LabelStack;
@@ -72,13 +74,19 @@ public class LabelAccumulator extends ColumnOverrideLabelAccumulator {
     }
 
     int columnIndex = bodyDataLayer.getColumnIndexByPosition(columnPosition);
-    ColumnType columnType = bodyDataProvider.getColumns().get(columnIndex).getColumnType();
-    if (columnType == ColumnType.SPAN_ANNOTATION) {
-      // Assign the span annotation style label to cells in columns with the respective flag.
-      configLabels.addLabel(StyleConfiguration.SPAN_ANNOTATION_CELL_STYLE);
-    } else if (columnType == ColumnType.TOKEN_TEXT) {
-      // Assign the token text style label to cells in columns with the respective flag.
-      configLabels.addLabel(StyleConfiguration.TOKEN_TEXT_CELL_STYLE);
+    List<Column> columns = bodyDataProvider.getColumns();
+    ColumnType columnType = null;
+    if (columns.size() != 0) {
+      columnType = columns.get(columnIndex).getColumnType();
+    }
+    if (columnType != null) {
+      if (columnType == ColumnType.SPAN_ANNOTATION) {
+        // Assign the span annotation style label to cells in columns with the respective flag.
+        configLabels.addLabel(StyleConfiguration.SPAN_ANNOTATION_CELL_STYLE);
+      } else if (columnType == ColumnType.TOKEN_TEXT) {
+        // Assign the token text style label to cells in columns with the respective flag.
+        configLabels.addLabel(StyleConfiguration.TOKEN_TEXT_CELL_STYLE);
+      }
     }
   }
 

--- a/docs/dev/src/development/automated-tests/ui-integration-tests.md
+++ b/docs/dev/src/development/automated-tests/ui-integration-tests.md
@@ -65,3 +65,12 @@ You can set the execution order within a test class explicitly with the `@Order(
 If you integrate a new bundle that has not been tested before, you have to add the bundle manually to the 
 dependencies of `org.corpus_tools.hexatomic.it.tests`.
 Adding it to the feature or product is not enough.
+
+Note that the Hexatomic integration tests use the SWTBot *EN_US* keyboard layout.
+If you see UI tests failing on your local machine, you may have to install US English language support.
+In other words, the *locale en_US* must be available on your operating system.
+
+On Linux, you can test this with the command `locale -a`.
+If `en_US.utf8` is listed in the output, you're all set up.
+
+For more information on SWTBot's keyboard layouts, see [Keyboard Layouts in SWTBot on the Eclipse wiki](https://wiki.eclipse.org/SWTBot/Keyboard_Layouts).

--- a/tests/org.corpus_tools.hexatomic.it.tests/src/main/java/org/corpus_tools/hexatomic/it/tests/TestCorpusStructure.java
+++ b/tests/org.corpus_tools.hexatomic.it.tests/src/main/java/org/corpus_tools/hexatomic/it/tests/TestCorpusStructure.java
@@ -41,6 +41,7 @@ class TestCorpusStructure {
   void setup() {
     org.eclipse.swtbot.swt.finder.utils.SWTBotPreferences.KEYBOARD_STRATEGY =
         "org.eclipse.swtbot.swt.finder.keyboard.SWTKeyboardStrategy";
+    org.eclipse.swtbot.swt.finder.utils.SWTBotPreferences.KEYBOARD_LAYOUT = "EN_US";
 
     IEclipseContext ctx = ContextHelper.getEclipseContext();
 

--- a/tests/org.corpus_tools.hexatomic.it.tests/src/main/java/org/corpus_tools/hexatomic/it/tests/TestCorpusStructure.java
+++ b/tests/org.corpus_tools.hexatomic.it.tests/src/main/java/org/corpus_tools/hexatomic/it/tests/TestCorpusStructure.java
@@ -29,7 +29,7 @@ import org.junit.jupiter.api.TestMethodOrder;
 @TestMethodOrder(OrderAnnotation.class)
 class TestCorpusStructure {
 
-  private SWTWorkbenchBot bot = new SWTWorkbenchBot(ContextHelper.getEclipseContext());
+  private SWTWorkbenchBot bot = new SWTWorkbenchBot(TestHelper.getEclipseContext());
 
   private ECommandService commandService;
   private EHandlerService handlerService;
@@ -41,9 +41,9 @@ class TestCorpusStructure {
   void setup() {
     org.eclipse.swtbot.swt.finder.utils.SWTBotPreferences.KEYBOARD_STRATEGY =
         "org.eclipse.swtbot.swt.finder.keyboard.SWTKeyboardStrategy";
-    org.eclipse.swtbot.swt.finder.utils.SWTBotPreferences.KEYBOARD_LAYOUT = "EN_US";
+    TestHelper.setKeyboardLayout();
 
-    IEclipseContext ctx = ContextHelper.getEclipseContext();
+    IEclipseContext ctx = TestHelper.getEclipseContext();
 
     projectManager = ContextInjectionFactory.make(ProjectManager.class, ctx);
 

--- a/tests/org.corpus_tools.hexatomic.it.tests/src/main/java/org/corpus_tools/hexatomic/it/tests/TestGraphEditor.java
+++ b/tests/org.corpus_tools.hexatomic.it.tests/src/main/java/org/corpus_tools/hexatomic/it/tests/TestGraphEditor.java
@@ -56,6 +56,7 @@ class TestGraphEditor {
   void setup() {
     org.eclipse.swtbot.swt.finder.utils.SWTBotPreferences.KEYBOARD_STRATEGY =
         "org.eclipse.swtbot.swt.finder.keyboard.SWTKeyboardStrategy";
+    org.eclipse.swtbot.swt.finder.utils.SWTBotPreferences.KEYBOARD_LAYOUT = "EN_US";
 
     IEclipseContext ctx = ContextHelper.getEclipseContext();
 

--- a/tests/org.corpus_tools.hexatomic.it.tests/src/main/java/org/corpus_tools/hexatomic/it/tests/TestGraphEditor.java
+++ b/tests/org.corpus_tools.hexatomic.it.tests/src/main/java/org/corpus_tools/hexatomic/it/tests/TestGraphEditor.java
@@ -42,7 +42,7 @@ import org.junit.jupiter.api.TestMethodOrder;
 @TestMethodOrder(OrderAnnotation.class)
 class TestGraphEditor {
 
-  private final SWTWorkbenchBot bot = new SWTWorkbenchBot(ContextHelper.getEclipseContext());
+  private final SWTWorkbenchBot bot = new SWTWorkbenchBot(TestHelper.getEclipseContext());
 
   private URI exampleProjectUri;
   private ECommandService commandService;
@@ -56,9 +56,9 @@ class TestGraphEditor {
   void setup() {
     org.eclipse.swtbot.swt.finder.utils.SWTBotPreferences.KEYBOARD_STRATEGY =
         "org.eclipse.swtbot.swt.finder.keyboard.SWTKeyboardStrategy";
-    org.eclipse.swtbot.swt.finder.utils.SWTBotPreferences.KEYBOARD_LAYOUT = "EN_US";
+    TestHelper.setKeyboardLayout();
 
-    IEclipseContext ctx = ContextHelper.getEclipseContext();
+    IEclipseContext ctx = TestHelper.getEclipseContext();
 
     errorService = ContextInjectionFactory.make(ErrorService.class, ctx);
     projectManager = ContextInjectionFactory.make(ProjectManager.class, ctx);

--- a/tests/org.corpus_tools.hexatomic.it.tests/src/main/java/org/corpus_tools/hexatomic/it/tests/TestGridEditor.java
+++ b/tests/org.corpus_tools.hexatomic.it.tests/src/main/java/org/corpus_tools/hexatomic/it/tests/TestGridEditor.java
@@ -52,7 +52,7 @@ public class TestGridEditor {
 
   private static final String TEST_ANNOTATION_VALUE = "TEST";
 
-  private SWTWorkbenchBot bot = new SWTWorkbenchBot(ContextHelper.getEclipseContext());
+  private SWTWorkbenchBot bot = new SWTWorkbenchBot(TestHelper.getEclipseContext());
 
   private URI exampleProjectUri;
   private URI overlappingExampleProjectUri;
@@ -67,9 +67,9 @@ public class TestGridEditor {
 
   @BeforeEach
   void setup() {
-    org.eclipse.swtbot.swt.finder.utils.SWTBotPreferences.KEYBOARD_LAYOUT = "EN_US";
+    TestHelper.setKeyboardLayout();
 
-    IEclipseContext ctx = ContextHelper.getEclipseContext();
+    IEclipseContext ctx = TestHelper.getEclipseContext();
 
     ctx.set(ErrorService.class, errorService);
 

--- a/tests/org.corpus_tools.hexatomic.it.tests/src/main/java/org/corpus_tools/hexatomic/it/tests/TestGridEditor.java
+++ b/tests/org.corpus_tools.hexatomic.it.tests/src/main/java/org/corpus_tools/hexatomic/it/tests/TestGridEditor.java
@@ -67,6 +67,8 @@ public class TestGridEditor {
 
   @BeforeEach
   void setup() {
+    org.eclipse.swtbot.swt.finder.utils.SWTBotPreferences.KEYBOARD_LAYOUT = "EN_US";
+
     IEclipseContext ctx = ContextHelper.getEclipseContext();
 
     ctx.set(ErrorService.class, errorService);

--- a/tests/org.corpus_tools.hexatomic.it.tests/src/main/java/org/corpus_tools/hexatomic/it/tests/TestHelper.java
+++ b/tests/org.corpus_tools.hexatomic.it.tests/src/main/java/org/corpus_tools/hexatomic/it/tests/TestHelper.java
@@ -8,22 +8,32 @@ import org.osgi.framework.Bundle;
 import org.osgi.framework.FrameworkUtil;
 
 /**
- * A helper class to get an Eclipse context in test to create {@link SWTWorkbenchBot} instances.
+ * Provides static helper methods.
+ * 
  * @author Thomas Krause
+ * @author Stephan Druskat
  *
  */
-public class ContextHelper {
-  
+public class TestHelper {
+
   /**
-   * Get the Eclipse context for testing.
+   * Get the Eclipse context for testing to create {@link SWTWorkbenchBot} instances.
+   * 
    * @return The context.
    */
   public static IEclipseContext getEclipseContext() {
     Bundle activatorBundle = FrameworkUtil.getBundle(Activator.class);
-    final IEclipseContext serviceContext = EclipseContextFactory
-        .getServiceContext(activatorBundle.getBundleContext());
+    final IEclipseContext serviceContext =
+        EclipseContextFactory.getServiceContext(activatorBundle.getBundleContext());
 
     return serviceContext.get(IWorkbench.class).getApplication().getContext();
+  }
+
+  /**
+   * Sets the SWTBot keyboard layout to <code>EN_US</code>.
+   */
+  public synchronized static void setKeyboardLayout() {
+    org.eclipse.swtbot.swt.finder.utils.SWTBotPreferences.KEYBOARD_LAYOUT = "EN_US";
   }
 
 }

--- a/tests/org.corpus_tools.hexatomic.it.tests/src/main/java/org/corpus_tools/hexatomic/it/tests/TestHelper.java
+++ b/tests/org.corpus_tools.hexatomic.it.tests/src/main/java/org/corpus_tools/hexatomic/it/tests/TestHelper.java
@@ -16,6 +16,10 @@ import org.osgi.framework.FrameworkUtil;
  */
 public class TestHelper {
 
+  private TestHelper() {
+    // Hide default constructor
+  }
+
   /**
    * Get the Eclipse context for testing to create {@link SWTWorkbenchBot} instances.
    * 

--- a/tests/org.corpus_tools.hexatomic.it.tests/src/main/java/org/corpus_tools/hexatomic/it/tests/TestHelper.java
+++ b/tests/org.corpus_tools.hexatomic.it.tests/src/main/java/org/corpus_tools/hexatomic/it/tests/TestHelper.java
@@ -32,7 +32,7 @@ public class TestHelper {
   /**
    * Sets the SWTBot keyboard layout to <code>EN_US</code>.
    */
-  public synchronized static void setKeyboardLayout() {
+  public static void setKeyboardLayout() {
     org.eclipse.swtbot.swt.finder.utils.SWTBotPreferences.KEYBOARD_LAYOUT = "EN_US";
   }
 

--- a/tests/org.corpus_tools.hexatomic.it.tests/src/main/java/org/corpus_tools/hexatomic/it/tests/TestProjectManager.java
+++ b/tests/org.corpus_tools.hexatomic.it.tests/src/main/java/org/corpus_tools/hexatomic/it/tests/TestProjectManager.java
@@ -52,7 +52,7 @@ class TestProjectManager {
 
   @BeforeEach
   void setup() {
-    IEclipseContext ctx = ContextHelper.getEclipseContext();
+    IEclipseContext ctx = TestHelper.getEclipseContext();
 
     bot = new SWTWorkbenchBot(ctx);
 

